crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -17,244 +20,237 @@ use crate::other::arguments::{
+        AnyNodeRef::Arguments(arguments) => {
+            handle_arguments_separator_comment(comment, locator, arguments)
+        }
+        AnyNodeRef::Comprehension(comprehension) => {
+            handle_comprehension_comment(comment, locator, comprehension)
+        }
+        AnyNodeRef::ExprAttribute(attribute) => handle_attribute_comment(comment, attribute),
+        AnyNodeRef::ExprBinOp(binary_expression) => {
+            handle_trailing_binary_expression_left_or_operator_comment(
+                comment,
+                locator,
+                binary_expression,
+            )
+        }
+        AnyNodeRef::ExprDict(_) | AnyNodeRef::Keyword(_) => {
+            handle_dict_unpacking_comment(comment, locator)
+        }
+        AnyNodeRef::ExprIfExp(expr_if) => handle_expr_if_comment(comment, locator, expr_if),
+        AnyNodeRef::ExprSlice(expr_slice) => handle_slice_comments(comment, locator, expr_slice),
+        AnyNodeRef::ExprStarred(starred) => {
+            handle_trailing_expression_starred_star_end_of_line_comment(comment, starred)
+        }
+        AnyNodeRef::ExprSubscript(expr_subscript) => {
+            if let Expr::Slice(expr_slice) = expr_subscript.slice.as_ref() {
+                handle_slice_comments(comment, locator, expr_slice)
+            } else {
+                CommentPlacement::Default(comment)
+            }
         }
+        AnyNodeRef::MatchCase(match_case) => handle_match_comment(comment, locator, match_case),
+        AnyNodeRef::ModModule(_) => {
+            handle_module_level_own_line_comment_before_class_or_function_comment(comment, locator)
+        }
+        AnyNodeRef::WithItem(_) => handle_with_item_comment(comment, locator),
+        AnyNodeRef::StmtFunctionDef(_) | AnyNodeRef::StmtAsyncFunctionDef(_) => {
+            handle_leading_function_with_decorators_comment(comment)
+        }
+        _ => CommentPlacement::Default(comment),
     }
 }
 
-/// Handles comments between except handlers and between the last except handler and any following `else` or `finally` block.
-fn handle_in_between_except_handlers_or_except_handler_and_else_or_finally_comment<'a>(
+fn handle_end_of_line_comment_around_body<'a>(
     comment: DecoratedComment<'a>,
     locator: &Locator,
 ) -> CommentPlacement<'a> {
-    if comment.line_position().is_end_of_line() {
-        return CommentPlacement::Default(comment);
+    // Handle comments before the first statement in a body
+    // ```python
+    // for x in range(10): # in the main body ...
+    //     pass
+    // else: # ... and in alternative bodies
+    //     pass
+    // ```
+    if let Some(following) = comment.following_node() {
+        if is_first_statement_in_body(following, comment.enclosing_node()) {
+            if SimpleTokenizer::new(
+                locator.contents(),
+                TextRange::new(comment.end(), following.start()),
+            )
+            .skip_trivia()
+            .next()
+            .is_none()
```

Comment:
Nit: The `if` is hard to read because of the formatting. It is also a rather complicated condition (I had to think for a while what it means if next is none). That's why I think assigning the result to a variable with a descriptive name will help future readers.